include/modem/location.h

```diff
@@ -121,17 +162,40 @@ struct location_data {
 	float accuracy;
 	/** Date and time (UTC). */
 	struct location_datetime datetime;
+
+#if defined(CONFIG_LOCATION_DATA_DETAILS)
+	/** Location details. */
+	struct location_data_details details;
+#endif
 };
 
+#if defined(CONFIG_LOCATION_DATA_DETAILS)
+/** Location error information. */
+struct location_data_error {
+	/** Data details at the time of error. */
+	struct location_data_details details;
+};
+#endif
+
 /** Location event data. */
 struct location_event_data {
 	/** Event ID. */
 	enum location_event_id id;
+	/** Used location method. */
+	enum location_method method;
 
 	union {
 		/** Current location, used with event LOCATION_EVT_LOCATION. */
 		struct location_data location;
 
+#if defined(CONFIG_LOCATION_DATA_DETAILS)
+		/**
+		 * Relevant location data when a timeout or an error occurs.
+		 * Used with event LOCATION_EVT_TIMEOUT and LOCATION_EVT_ERROR.
+		 */
+		struct location_data_error error;
+#endif
```

Comment:
This PR got merged now that CI was fine and necessary approvals in place. I get conflicts regularly so wanted to close this as there were no real code changes requested anymore. If this conversation brings any changes, I can still add them into a follow-up PR which is going to occur anyway.

Generally about data details. This has been a feature we've talked about for a year or so. It's been often called metrics etc. We have plans for some more parameters like error string, maybe error code, search time, etc. However, as those things are not finalized, I added simply what we needed for ATv2, and within the PVT there is some extra information.
So overall data metrics, or details or whatever we want to call it has information for both failing (timeout or error) and successful cases.

> Guarding error data behind `CONFIG_LOCATION_DATA_DETAILS` feels off to me. I would add a comment stating that "at the moment, location_data_error only contains location_data_details".

I didn't add this text yet (but can do next week if we get to that conclusion).
Normally application can handle errors and timeouts without additional information. But if they wish to receive more details about why the failure occurred, `location_data_error` has more information. And hence it could be flagged with `CONFIG_LOCATION_DATA_DETAILS`. I'm probably not completely following why this is such a bad thing.

> To get around this, consider having `location_data_details` be, instead, `location_event_details` and be a member directly of `location_event_data`, rather than having it live inside .error and .location

We've had few iterations of data details. I made few implementations and one version was with `location_data_details` being a direct member of `location_event_data`. A draft version is still seen here: https://github.com/trantanen/sdk-nrf/commit/23068e449ccd444c77497fa480cde9a6e0aa0b2a

The drawbacks of this approach are:
* You need to tell when these details are valid, i.e., for which events, and maybe document somewhere that there is already some information within the union.
* Application may need to store both `location_data_details` and `location_data`. Yep, you could store the entire `location_event_data`

Current approach is cleaner from the API user point of view as it can get one member of the union for a given location event and that's it.

> This would get rid of the somewhat confusing guard, allow you to remove the location_data_error type altogether, and simply the location details setter logic

We made a conscious decision to have this mentioned tweak in the location details setter logic, because we felt that the API would be nicer to use for the client the way it's defined now. We couldn't figure out a perfect solution so we had to make a compromise anyway.

Reply:
> Normally application can handle errors and timeouts without additional information. But if they wish to receive more details about why the failure occurred, location_data_error has more information. And hence it could be flagged with CONFIG_LOCATION_DATA_DETAILS. I'm probably not completely following why this is such a bad thing.

I don't consider that a bad thing, what bothers me more is that `.error` dissappears if you don't have details enabled. That's counter-intuitive, why would "details" disable "error"? It makes sense once you realize the only thing `.error` contains is details, but then, why not just have details directly in the event struct, rather than the location data struct?


> Application may need to store both location_data_details and location_data. Yep, you could store the entire location_event_data

> Current approach is cleaner from the API user point of view as it can get one member of the union for a given location event and that's it.

You also moved .method from .location to be directly in the event struct, didn't you? MMSS, for instance, now has to pass the whole event struct around rather than the .location component. Other code that needs both the details and the location data could do the same, right?


> You need to tell when these details are valid, i.e., for which events, and maybe document somewhere that there is already some information within the union.

I think you need to do this either way, right? Currently, to access the .error member, you have to check whether the event is an error. I think it would be fine to have the convention be that if the event type is an error, extra information will be present, or something to that effect